### PR TITLE
Blackboxai/withdraw rejected submission

### LINKED ISSUE #1722: Add Submission Withdrawal Window After Rejection
## Overview
When a submission is rejected, the submitter has no way to withdraw and re-submit with a corrected proof. The current model is final-rejection with no appeal for proof correction.

### Features:
- withdraw_submission(env, caller, quest_id) allowing submitter to retract a rejected submission
- Re-submission allowed within quest deadline after withdrawal
- Emit SubmissionWithdrawn event

### Tasks:
- Add withdraw_submission in submission.rs`n- Update state machine to allow Rejected -> Withdrawn -> Pending transition
- Add tests

### Impacted Files:
- contracts/earn-quest/src/submission.rs`n
## Acceptance Criteria:
- Submitter can withdraw a rejected submission
- Re-submission creates a new pending entry

### PR DESCRIPTION
**Pull request by Meet-hybrid**
Summary

Closes #1722 
Adds an “withdraw rejected submission” flow so users can correct their proof and re-submit within the quest deadline. The contract now supports:

withdraw_submission(env, caller, quest_id) for the submitter
state transition: Rejected -> Withdrawn -> Pending
SubmissionWithdrawn event emission
re-submission after withdrawal
Motivation / Problem
Currently, once a submission is rejected there is no way to withdraw it and re-submit corrected proof, making the workflow “final-rejection” with no appeal/correction path.

Changes made
New submission state
SubmissionStatus::Withdrawn added
State machine update
Allowed transitions updated to include:
Rejected -> Withdrawn
Withdrawn -> Pending
New event
Added submission_withdrawn publisher with indexed topics (quest_id, submitter)
Event topic name: sub_wd
New contract entrypoint
withdraw_submission(env, quest_id, caller)
Requires:
caller auth (caller.require_auth())
submission status must be Rejected
quest must not be expired
Updates submission status to Withdrawn
Emits SubmissionWithdrawn
Re-submission logic
commit_submission and submit_proof now allow overwriting an existing submission only if it is in Withdrawn state.
Tests
Added contracts/earn-quest/tests/test_withdraw_submission.rs covering:
withdraw fails unless status is Rejected
withdraw sets status to Withdrawn and emits event
re-submission after withdrawal sets status to Pending
Files modified / added
contracts/earn-quest/src/types.rs
contracts/earn-quest/src/validation.rs
contracts/earn-quest/src/events.rs
contracts/earn-quest/src/errors.rs
contracts/earn-quest/src/submission.rs
contracts/earn-quest/src/lib.rs
contracts/earn-quest/tests/test_withdraw_submission.rs
contracts/earn-quest/src/withdraw_submission_note.md emptied (temporary)

Acceptance Criteria
✅ Submitter can withdraw a rejected submission
✅ Re-submission after withdrawal creates a new Pending submission (overwrite under same (quest_id, submitter) key)

### CODE REVIEW

File: contracts/earn-quest/src/submission.rs [hunk 1]
```diff
@@ -101,7 +105,7 @@ pub fn reveal_submission(
         return Err(Error::InvalidCommitment);
```
Comment (RUKAYAT-CODER) [inline contracts/earn-quest/src/submission.rs:105]: This second return is unreachable and should be removed. keep one one appropriate error

File: contracts/earn-quest/src/submission.rs [hunk 2]
```diff
@@ -182,6 +197,36 @@ pub fn submit_proof(
...
+    submitter: &Address,
+) -> Result<(), Error> {
+    submitter.require_auth();
+
+
+    let quest = storage::get_quest(env, quest_id)?;
+    validation::validate_quest_not_expired(env, quest.deadline)?;
+
+    let mut submission = storage::get_submission(env, quest_id, submitter)?;
+
```
Comment (RUKAYAT-CODER) [inline contracts/earn-quest/src/submission.rs:211]: add this comment
// Transition: Rejected -> Withdrawn (allows future re-submission)

**Review by RUKAYAT-CODER — COMMENTED** [review]
This is a solid implementation the feature is correctly implemented and aligns well with the specification. The state transition from Rejected, Withdrawn, Pending is handled properly, and the submission flow behaves as expected.
- You have duplicate auth check at lib.rs
caller.require_auth();
and in submission.rs
submitter.require_auth();
Just keep one recommended or best to be placed in lib.rs(entrypoint layer)
So in submission.rs, REMOVE:
- Add missing deadline edge case
File:
contracts/earn-quest/tests/test_withdraw_submission.rs
Also you can improve your code with little comments
